File: samples/java/Bingo/Server/Play/src/main/java/systems/zlink/samples/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/handlers/SubmitBingoCardHandler.java
package systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers;

import systems.zlink.framework.ZLinkMessageContext;
import systems.zlink.framework.spots.ZLinkSpotActorRequestHandler;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.actors.PlayerActor;
import systems.zlink.samples.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.BingoRoomSpot;
import systems.zlink.samples.bingo.shared.contracts.Messages;

import java.util.concurrent.CompletableFuture;
import java.util.concurrent.CompletionStage;

public final class SubmitBingoCardHandler
        implements ZLinkSpotActorRequestHandler<
                BingoRoomSpot,
                PlayerActor,
                Messages.SubmitBingoCardReq,
                Messages.SubmitBingoCardRes> {
    @Override
    public CompletionStage<Messages.SubmitBingoCardRes> handle(
            BingoRoomSpot spot,
            PlayerActor actor,
            ZLinkMessageContext context,
            Messages.SubmitBingoCardReq request) {
        return CompletableFuture.completedFuture(spot.submitCard(actor, request));
    }
}

File: samples/kotlin/Bingo/Server/Play/src/main/kotlin/systems/zlink/samples/kotlin/bingo/server/play/infrastructure/zlink/spots/bingoroomspot/handlers/SubmitBingoCardHandler.kt
package systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.handlers

import systems.zlink.framework.ZLinkMessageContext
import systems.zlink.framework.kotlin.ZLinkSuspendingSpotActorRequestHandler
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.actors.PlayerActor
import systems.zlink.samples.kotlin.bingo.server.play.infrastructure.zlink.spots.bingoroomspot.BingoRoomSpot
import systems.zlink.samples.kotlin.bingo.shared.contracts.SubmitBingoCardReq
import systems.zlink.samples.kotlin.bingo.shared.contracts.SubmitBingoCardRes

class SubmitBingoCardHandler() :
    ZLinkSuspendingSpotActorRequestHandler<
        BingoRoomSpot,
        PlayerActor,
        SubmitBingoCardReq,
        SubmitBingoCardRes,
    > {
    override suspend fun handle(
        spot: BingoRoomSpot,
        actor: PlayerActor,
        context: ZLinkMessageContext,
        request: SubmitBingoCardReq,
    ): SubmitBingoCardRes = spot.submitCard(actor, request)
}
